### DISCUSSION TITLE
postgresql goes into readinessprobe livenessprobe failure loop

### DISCUSSION THREAD

**Discussion by yongjunkim42** [apache/airflow#37747] [Q&A]
Hi everyone, I am trying to install airflow using helm like the command below: 

	helm install airflow apache-airflow/airflow --debug  \
	--version "1.8.0" \
	-n $(namespace) \
	--values custom_values_local_development.yaml \

However, this attempt fails as postgresql pod repeatedly fails readiness and liveness probes, and go into a reset loop.

    Events:
      Type     Reason     Age               From               Message
      ----     ------     ----              ----               -------
      Normal   Scheduled  46s               default-scheduler  Successfully assigned airflow2/airflow-postgresql-0 to minikube
      Normal   Pulled     43s               kubelet            Container image "docker.io/bitnami/postgresql:11" already present on machine
      Normal   Created    43s               kubelet            Created container postgresql
      Normal   Started    43s               kubelet            Started container postgresql
      Warning  Unhealthy  6s (x4 over 36s)  kubelet            Readiness probe failed: 127.0.0.1:5432 - no attempt
      Warning  Unhealthy  6s                kubelet            Liveness probe failed: 127.0.0.1:5432 - no attempt

This seems like bit odd, so I took a look at the pod configuration, and discovered the initial delay on the postgresql is rather short

      Liveness:   exec [/bin/sh -c exec pg_isready -U "" -h 127.0.0.1 -p 5432] delay=30s timeout=5s period=10s #success=1 #failure=6
      Readiness:  exec [/bin/sh -c -e exec pg_isready -U "" -h 127.0.0.1 -p 5432 [ -f /opt/bitnami/postgresql/tmp/.initialized ] || [ -f /bitnami/postgresql/.initialized ]] delay=5s timeout=5s period=10s #success=1 #failure=6

To fix this, I tried to add livenessProbe and readinessProbe parameters on my values yaml file(custom_values_local_development.yaml) like below, but this update does not seem to pass to the postgresql pod:

    postgresql:
      livenessProbe:
        initialDelaySeconds: 300
      readinessProbe:
        initialDelaySeconds: 300
      enabled: true
      auth:
        postgresqlPassword: postgres
        postgresqlUsername: postgres
      postgresqlExtendedConf:
        log_error_verbosity: verbose

However, this does not seem to work, so I checked out the default values yaml file, and it seems like livenessProbe or readinessProbe are not the parameters I can set for postgresql pod portion unlike webserver pod?  Here is the existing values.yaml on the repo:

    postgresql:
      enabled: true
      auth:
        enablePostgresUser: true
        postgresPassword: postgres
        username: ""
        password: ""
    ...
    webserver:
      enabled: true
      # Add custom annotations to the webserver configmap
      configMapAnnotations: {}
      #  hostAliases for the webserver pod
      hostAliases: []
      #  - ip: "127.0.0.1"
      #    hostnames:
      #      - "foo.local"
      #  - ip: "10.1.2.3"
      #    hostnames:
      #      - "foo.remote"
      allowPodLogReading: true
      livenessProbe:
        initialDelaySeconds: 15
        timeoutSeconds: 5
        failureThreshold: 5
        periodSeconds: 10
        scheme: HTTP
      readinessProbe:
        initialDelaySeconds: 15
        timeoutSeconds: 5
        failureThreshold: 5
        periodSeconds: 10
        scheme: HTTP

Now, I am quite lost as I know postgresql pod is failing due to fast readiness or liveness probes, but can't really fix it through values.yaml.  Am I looking into a dead end, and perhaps the issue lies in somewhere else?

**Comment by potiuk** [comment]
The postgres we are using is from bitnami -it's a subchart of airflow's chart and you should look there.  And it's definitely discourage for use in production - that's why it does not have almost any controls. You should rather consider switching to "regular" database managed outside of the chart.

I think also that you have bigger problem (like resources) not the liveness probe timeout you need find out why it's not starting  - in the usual  way you check it in your k8s instance.

**Reply by yongjunkim42** [reply]
I will venture to bitnami side as well, and scan through any similar issues, and ask questions.  Thanks for the pointer.

For the purpose of this DB, this is actually for local testing and not for production.  Will definitely consider using a different database for it.

For the bigger problems you mention, do you have a usual list of things I should be checking on?  Here is the "describe" on the postgreSQL pod, and I am going through the relevant bitnami discussions on this issue, but not finding where I should focus on

    Name:             airflow-postgresql-0
    Namespace:        airflow2
    Priority:         0
    Service Account:  default
    Node:             minikube/192.168.49.2
    Start Time:       Wed, 28 Feb 2024 10:41:22 -0500
    Labels:           app.kubernetes.io/component=primary
                      app.kubernetes.io/instance=airflow
                      app.kubernetes.io/managed-by=Helm
                      app.kubernetes.io/name=postgresql
                      controller-revision-hash=airflow-postgresql-6fb7fb4698
                      helm.sh/chart=postgresql-12.1.9
                      statefulset.kubernetes.io/pod-name=airflow-postgresql-0
    Annotations:      <none>
    Status:           Running
    IP:               172.17.0.3
    IPs:
      IP:           172.17.0.3
    Controlled By:  StatefulSet/airflow-postgresql
    Containers:
      postgresql:
        Container ID:   docker://c1e3c27d4e220b99156b0370bc8825e934fafc3a8e203430a7b14f0d94002ba3
        Image:          docker.io/bitnami/postgresql:11
        Image ID:       docker-pullable://bitnami/postgresql@sha256:93bf18bd5a825c03f3054abdab5d3227a72692a4ce6ed430c60354ccb9ccc8ff
        Port:           5432/TCP
        Host Port:      0/TCP
        State:          Running
          Started:      Wed, 28 Feb 2024 11:02:43 -0500
        Last State:     Terminated
          Reason:       Completed
          Exit Code:    0
          Started:      Wed, 28 Feb 2024 11:01:18 -0500
          Finished:     Wed, 28 Feb 2024 11:02:42 -0500
        Ready:          False
        Restart Count:  9
        Requests:
          cpu:      250m
          memory:   256Mi
        Liveness:   exec [/bin/sh -c exec pg_isready -U "" -h 127.0.0.1 -p 5432] delay=30s timeout=5s period=10s #success=1 #failure=6
        Readiness:  exec [/bin/sh -c -e exec pg_isready -U "" -h 127.0.0.1 -p 5432
    [ -f /opt/bitnami/postgresql/tmp/.initialized ] || [ -f /bitnami/postgresql/.initialized ]
    ] delay=5s timeout=5s period=10s #success=1 #failure=6
        Environment:
          BITNAMI_DEBUG:                        false
          POSTGRESQL_PORT_NUMBER:               5432
          POSTGRESQL_VOLUME_DIR:                /bitnami/postgresql
          PGDATA:                               /bitnami/postgresql/data
           ...[Account Info Deleted]
          POSTGRESQL_ENABLE_LDAP:               no
          POSTGRESQL_ENABLE_TLS:                no
          POSTGRESQL_LOG_HOSTNAME:              false
          POSTGRESQL_LOG_CONNECTIONS:           false
          POSTGRESQL_LOG_DISCONNECTIONS:        false
          POSTGRESQL_PGAUDIT_LOG_CATALOG:       off
          POSTGRESQL_CLIENT_MIN_MESSAGES:       error
          POSTGRESQL_SHARED_PRELOAD_LIBRARIES:  pgaudit
        Mounts:
          /bitnami/postgresql from data (rw)
          /dev/shm from dshm (rw)
          /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-sjhgh (ro)
    Conditions:
      Type              Status
      Initialized       True 
      Ready             False 
      ContainersReady   False 
      PodScheduled      True 
    Volumes:
      data:
        Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
        ClaimName:  data-airflow-postgresql-0
        ReadOnly:   false
      dshm:
        Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
        Medium:     Memory
        SizeLimit:  <unset>
      kube-api-access-sjhgh:
        Type:                    Projected (a volume that contains injected data from multiple sources)
        TokenExpirationSeconds:  3607
        ConfigMapName:           kube-root-ca.crt
        ConfigMapOptional:       <nil>
        DownwardAPI:             true
    QoS Class:                   Burstable
    Node-Selectors:              <none>
    Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                                 node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
    

If I should be concerned about the resource allocation, maybe the requested capacity is too small at 
    cpu:      250m
    memory:   256Mi

However, I am not sure what should be appropriate amounts for the DB.  THanks!

**Reply by potiuk** [reply]
Nope. No list of things, assumption is that those who choose k8s generally do it because they are used to use k8s and now how to check if thigs are not working. 

What I might say - that for development environment of airflow itself - default settings work when you deploy it on a local k9s with kind - for our development environment :https://github.com/apache/airflow/blob/main/contributing-docs/testing/k8s_tests.rst  and it's best if you have ~4 GB memory allocated for docker if you run it locally if you deploy k8s.

Generally speaking logs of postgres while starting should help you. I personally use k9s more often than not to debug things with k8s - it's really helpful.

But that's about all I can help at this stage.

**Reply by potiuk** [reply]
It's very much possible that simply your k8s has just too little resources . 

I just used absolutely default settings on my MacOS and here it is; 

<img width="1051" alt="Screenshot 2024-02-28 at 17 22 36" src="https://github.com/apache/airflow/assets/595491/4fa11538-b66b-4c6e-95b6-4a50cdc84f15">

**Reply by yongjunkim42** [reply]
Thanks this is a really helpful insight.  I have some experience in using k8s, but not an expert, and defeinitely can learn to find out why things go the way they do.   I am working on this minikube for this.  I used the following command to start that minikube:

    minikube start --mount-string $(shell realpath .):/airflow --mount --driver=docker --cpus 4 --memory 8192

Here how it looks on k9s, as the installation fails
![image](https://github.com/apache/airflow/assets/36705812/f6980654-9ecc-4589-8368-29635ca2abbc)

I will try adjusting the allocated resource.

**Reply by yongjunkim42** [reply]
OK, I after many days trying, I resolved the issue.  The issue is mainly due to I have M1 Mac Book Pro.  This causes many components to not work as intended.  Here is my recommendation as of 3/7/2024:

1. If you can, do not use minikube, this will lead to multiple issues with Apple silicon chips.  I recommend using Kubernetes engine on Docker Desktop at the moment.
2. Do not use Helm Chart 1.8.0.  I had success using 1.7.0, but you might want to try different versions.
3. Do add liveness and readiness probe on you webserver section on values.yaml.  Do set initialDelaySeconds to be 300, so they don't activate probes too early leading to failure.  Here is what my config looks like:

```
## configs for the web Pods' readiness probe
readinessProbe:
   initialDelaySeconds: 300
   periodSeconds: 10
   timeoutSeconds: 5
   failureThreshold: 6

## configs for the web Pods' liveness probe
 livenessProbe:
   initialDelaySeconds: 300
   periodSeconds: 10
   timeoutSeconds: 5
   failureThreshold: 6
```

If all this does not work, I recommend using a different hardware setting or even a virtual environment like EC2 or Amazon Work Space instance.

**Comment by yongjunkim42** [comment]
Since the issue has been resolved as commented above..  I will close this discussion.
